EIPS/eip-2770.md

```diff
@@ -0,0 +1,178 @@
+`requestTypeHash` - hash of the registered relay request type
+`suffixData` - RLP-encoding of the remainder of the request struct
+`signature` - an ERC-712 signature on the concatenation of `forwardRequest` and `suffixData`
+
+#### Command execution
+
+In order for the Forwarder to perform an operation, the following method is to be called: 
+```
+function execute(
+   ForwardRequest forwardRequest,
+   bytes32 domainSeparator,
+   bytes32 requestTypeHash,
+   bytes suffixData,
+   bytes signature
+)
+public
+payable
+returns (
+   bool success,
+   bytes memory ret
+)
+```
+ 
+Performs the ‘verify’ internally and if it succeeds performs the following call:
+```
+bytes memory data = abi.encodePacked(forwardRequest.data, forwardRequest.from);
+(success, ret) = forwardRequest.to.call{gas: forwardRequest.gas, value: forwardRequest.value}(data);
+```
+Regardless of whether the inner call succeeds or reverts, the nonce is incremented, invalidating the signature and preventing a replay of the request.
+
+Note that `gas` parameter behaves according to EVM rules, specifically EIP-150.\
+In case there is not enough `value` in the Forwarder the execution reverts.\
+In case after the call there is a remaining Ether in the forwarder it is sent back to the signer. 
+ 
+#### ERC-712 and 'suffixData' parameter
+`suffixData` field must provide a valid 'tail' of an ERC-712 typed data.
+For instance, in order to sign on the `ExtendedRequest` struct, the data will be a concatenation of the following chunks:
+* `forwardRequest` fields will be RLP-encoded as-is, and variable-length `data` field will be hashed
+* `uint256 x` will be appended entirely as-is
+* `bytes z` will be hashed first
+* `ExtraData extraData` will be hashed as a typed data
+
+So a valid `suffixData` is calculated as following:
+```
+function calculateSuffixData(ExtendedRequest request) internal pure returns (bytes) {
+    return abi.encode(request.x, keccak256(request.z), hashExtraData(request.extraData));
+}
+
+function hashExtraData(ExtraData extraData) internal pure returns (bytes32) {
+    return keccak256(abi.encode(
+            keccak256("ExtraData(uint256 a,uint256 b,uint256 c)"),
+            extraData.a,
+            extraData.b,
+            extraData.c
+        ));
+}
+```
+
+#### Accepting Forwarded calls
+In order to support calls performed via the Forwarder, the Recipient contract must implement ERC - RECIPIENT.
+
+## Rationale
+Further relying on `msg.sender` to authenticate end users by their externally-owned accounts is taking the Ethereum dapp ecosystem to a dead end.
+
+A need for users to own Ether before they can interact with any contract has made a huge portion of use-cases for smart contracts non-viable,
+which in turn limits the mass adoption and enforces this vicious cycle.
+
+## Implementation
+The implementation can be found at https://github.com/opengsn/forwarder
```

Comment:
Implementations should either be inline or live in `../assets/eip-2770/<files>`, not be external links.  Alternatively, you can just leave this section out entirely as it is optional.